File: ur_script_generator/ur_script_generator/ur_script_generator.py
import os
import json
import math
import rclpy
from rclpy.node import Node
from ur_script_generator_msgs.srv import GenerateURScript
from tf_tools_msgs.srv import LookupTransform

BASEFRAME_ID = "base"
FACEPLATE_ID = "tool0"


class URScriptGenerator(Node):
    def __init__(self):
        super().__init__("ur_script_generator")
        self.srv = self.create_service(
            GenerateURScript, "ur_script_generator", self.ur_script_generator_callback
        )

        self.scene_parameters_path = self.declare_parameter(
            "scene_parameters_path", "default value"
        )
        self.scene_parameters_path = (
            self.get_parameter("scene_parameters_path")
            .get_parameter_value()
            .string_value
        )

        self.driver_parameters = self.declare_parameter(
            "driver_parameters", "default value"
        )

        self.driver_parameters = (
            self.get_parameter("driver_parameters").get_parameter_value().string_value
        )

        self.client = self.create_client(LookupTransform, "tf_lookup")
        self.request = LookupTransform.Request()
        self.response = LookupTransform.Response()

        while not self.client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info("tf lookup service not available, waiting again...")

        self.have_preferred_joint_configuration = False

    def ur_script_generator_callback(self, request, response):
        self.get_logger().info("Got request '{asdf}'".format(asdf=request))
        move_j = request.command == "move" and request.move_type == "move_j"
        move_l = request.command == "move" and request.move_type == "move_l"
        freedrive_mode = request.command == "freedrive_mode"
        end_freedrive_mode = request.command == "end_freedrive_mode"
        if move_j:
            response.ur_script = self.generate_move_j(request)
            return response
        elif move_l:
            response.ur_script = self.generate_move_l(request)
            return response
        elif freedrive_mode:
            response.ur_script = self.freedrive_mode(request)
            return response
        elif end_freedrive_mode:
            response.ur_script = self.end_freedrive_mode()
            return response
        else:
            self.get_logger().warn("Unknown request '{asdf}'".format(asdf=request))
            pass

    def tf_lookup(self, parent, child, deadline):
        self.request.parent_id = parent
        self.request.child_id = child
        self.request.deadline = deadline
        self.future = self.client.call_async(self.request)
        self.get_logger().info(f"TF lookup request sent: {self.request}.")
        while rclpy.ok():
            rclpy.spin_once(self)
            if self.future.done():
                try:
                    response = self.future.result()
                except Exception as e:
                    self.get_logger().error(
                        f"TF lookup service call failed with: {(e,)}."
                    )
                else:
                    self.response = response
                    self.get_logger().info(f"TF lookup result: {self.response}.")
                finally:
                    self.get_logger().info(f"TF lookup service call completed.")
                    return response

    def pose_to_string(self, tf_stamped):
        x = tf_stamped.transform.translation.x
        y = tf_stamped.transform.translation.y
        z = tf_stamped.transform.translation.z
        q = tf_stamped.transform.rotation
        angle = 2 * math.acos(q.w)
        den = math.sqrt(1 - q.w * q.w)
        if den < 0.001:
            Rx = q.x * angle
            Ry = q.y * angle
            Rz = q.z * angle
        else:
            Rx = (q.x / den) * angle
            Ry = (q.y / den) * angle
            Rz = (q.z / den) * angle

        return (
            "p["
            + str(x)
            + ", "
            + str(y)
            + ", "
            + str(z)
            + ", "
            + str(Rx)
            + ", "
            + str(Ry)
            + ", "
            + str(Rz)
            + "]"
        )

    def generate_move_j(self, request):
        self.get_logger().info(
            "Requested move_j to: %s with tool: %s"
            % (request.goal_feature_name, request.tcp_name)
        )

        pure_goal_feature_name = ""
        for i in range(9):
            if request.goal_feature_name.endswith("_" + str(i)):
                pure_goal_feature_name = request.goal_feature_name[:-2]

        if pure_goal_feature_name != "":
            self.goal_feature_params_path = os.path.join(
                self.scene_parameters_path,
                "frames",
                "{name}.json".format(name=pure_goal_feature_name),
            )
        else:
            self.goal_feature_params_path = os.path.join(
                self.scene_parameters_path,
                "frames",
                "{name}.json".format(name=request.goal_feature_name),
            )
        if self.goal_feature_params_path != None:
            try:
                with open(self.goal_feature_params_path) as f:
                    self.goal_feature_params = json.load(f)
                    if "preferred_joint_configuration" in self.goal_feature_params:
                        self.have_preferred_joint_configuration = True
                        j0 = self.goal_feature_params["preferred_joint_configuration"][
                            "j0"
                        ]
                        j1 = self.goal_feature_params["preferred_joint_configuration"][
                            "j1"
                        ]
                        j2 = self.goal_feature_params["preferred_joint_configuration"][
                            "j2"
                        ]
                        j3 = self.goal_feature_params["preferred_joint_configuration"][
                            "j3"
                        ]
                        j4 = self.goal_feature_params["preferred_joint_configuration"][
                            "j4"
                        ]
                        j5 = self.goal_feature_params["preferred_joint_configuration"][
                            "j5"
                        ]
                    else:
                        self.have_preferred_joint_configuration = False
            except Exception:
                self.have_preferred_joint_configuration = False
                self.get_logger().info("no json item")
            # self.goal_feature_params = json.load(open(self.goal_feature_params_path))

        else:
            self.have_preferred_joint_configuration = False
            j0 = 0.0
            j1 = 0.0
            j2 = 0.0
            j3 = 0.0
            j4 = 0.0
            j5 = 0.0

        target_in_base = self.tf_lookup(BASEFRAME_ID, request.goal_feature_name, 3000)

        if not target_in_base.success:
            self.get_logger().error("Falied to lookup target in base.")

        tcp_in_faceplate = self.tf_lookup(FACEPLATE_ID, request.tcp_name)
        if not tcp_in_faceplate.success:
            self.get_logger().error("Falied to lookup tcp in faceplate.")

        ur_script = "def ur_script():\n"
        ur_script += "  set_tcp(" + self.pose_to_string(tcp_in_faceplate) + ")\n"
        if self.have_preferred_joint_configuration:
            ur_script += (
                "  movej(get_inverse_kin("
                + self.pose_to_string(target_in_base)
                + ", ["
                + str(j0)
                + " , "
                + str(j1)
                + ", "
                + str(j2)
                + ", "
                + str(j3)
                + ", "
                + str(j4)
                + ", "
                + str(j5)
                + "])"
                + ",a="
                + str(request.acceleration)
                + ",v="
                + str(request.velocity)
                + ")\n"
            )
        else:
            ur_script += (
                "movej(get_inverse_kin("
                + self.pose_to_string(target_in_base)
                + "),a="
                + str(request.acceleration)
                + ",v="
                + str(request.velocity)
                + ")\n"
            )
        ur_script += "end\nur_script()"

        if target_in_base.success and tcp_in_faceplate.success:
            return ur_script
        else:
            return ""

    def generate_move_l(self, request):
        target_in_base = self.tf_lookup(BASEFRAME_ID, request.goal_feature_name, 3000)
        if not target_in_base.success:
            self.get_logger().error("Falied to lookup target in base.")
        tcp_in_faceplate = self.tf_lookup(FACEPLATE_ID, request.tcp_name)
        if not tcp_in_faceplate.success:
            self.get_logger().error("Falied to lookup tcp in faceplate.")
        ur_script = "def ur_script():\n"
        ur_script += "  set_tcp(" + self.pose_to_string(tcp_in_faceplate) + ")\n"
        ur_script += (
            "  movel("
            + self.pose_to_string(target_in_base)
            + ",a="
            + str(request.acceleration)
            + ",v="
            + str(request.velocity)
            + ")\n"
        )
        ur_script += "end\nur_script()"
        if target_in_base.success and tcp_in_faceplate.success:
            return ur_script
        else:
            return ""

    def generate_set_standard_digital_out(self, io, value):
        ur_script = "def ur_script():\n"
        ur_script += "  set_standard_digital_out(" + str(io) + "," + str(value) + ")\n"
        ur_script += "end\nur_script()"
        return ur_script

    def freedrive_mode(self):
        ur_script = "def ur_script():\n"
        ur_script += "  while True:\n"
        ur_script += "    freedrive_mode()\n"
        ur_script += "  end\n"
        ur_script += "end\nur_script()"
        return ur_script

    def end_freedrive_mode(self):
        ur_script = "def ur_script():\n"
        ur_script += "  end_freedrive_mode()\n"
        ur_script += "end\nur_script()"
        return ur_script


def main(args=None):
    rclpy.init(args=args)
    ursg = URScriptGenerator()
    rclpy.spin(ursg)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ur_script_generator/src/ur_script_generator.rs
use futures::stream::Stream;
use futures::StreamExt;
use r2r::ParameterValue;
use r2r::geometry_msgs::msg::TransformStamped;
use r2r::tf_tools_msgs::srv::LookupTransform;
use r2r::ur_tools_msgs::msg::JointPositions;
use r2r::ur_tools_msgs::msg::Payload;
use r2r::ur_tools_msgs::srv::GenerateURScript;
use r2r::ServiceRequest;
use serde::{Deserialize, Serialize};
use tera;

pub static BASEFRAME_ID: &'static str = "base";
pub static FACEPLATE_ID: &'static str = "tool0";

#[derive(Serialize, Deserialize)]
pub struct Interpretation {
    pub valid: bool,
    pub command: String,
    pub acceleration: f64,
    pub velocity: f64,
    pub use_execution_time: bool,
    pub execution_time: f32,
    pub use_blend_radius: bool,
    pub blend_radius: f32,
    pub use_joint_positions: bool,
    pub joint_positions: String,
    pub use_preferred_joint_config: bool,
    pub preferred_joint_config: String,
    pub use_payload: bool,
    pub payload: String,
    pub target_in_base: String,
    pub tcp_in_faceplate: String,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let ctx = r2r::Context::create()?;
    let mut node = r2r::Node::create(ctx, "ur_script_generator", "")?;

    let params = node.params.clone();
    let params_things =  params.lock().unwrap();
    let templates_path = match params_things.get("templates_path").unwrap() {
        ParameterValue::String(value) => value,
        _ => ""
    };

    let templates: tera::Tera = {
        let tera = match tera::Tera::new(&format!("{}/templates/*.script", templates_path)) {
            Ok(t) => {
                r2r::log_warn!(
                    "ur_script_generator",
                    "Searching for Tera templates, wait...",
                );
                t
            },
            Err(e) => {
                r2r::log_error!(
                    "ur_script_generator",
                    "UR Script template parsing error(s): {}", e
                ); 
                ::std::process::exit(1);
            }
        };
        tera
    };

    let service = node.create_service::<GenerateURScript::Service>("ur_script_generator")?;
    let tf_lookup_client = node.create_client::<LookupTransform::Service>("tf_lookup")?;
    let waiting_for_tf_lookup_server = node.is_available(&tf_lookup_client)?;

    let handle = std::thread::spawn(move || loop {
        node.spin_once(std::time::Duration::from_millis(100));
    });

    r2r::log_warn!(
        "ur_script_generator",
        "Waiting for TF Lookup service..."
    );
    waiting_for_tf_lookup_server.await?;
    r2r::log_info!(
        "ur_script_generator",
        "TF Lookup Service available."
    );
    r2r::log_info!(
        "ur_script_generator",
        "UR Script Generator Service node started."
    );

    tokio::task::spawn(async move {
        let result = ur_script_generator_server(service, tf_lookup_client, templates).await;
        match result {
            Ok(()) => r2r::log_info!(
                "ur_script_generator",
                "UR Script Generator Service call succeeded."
            ),
            Err(e) => r2r::log_error!(
                "ur_script_generator",
                "UR Script Generator Service call failed with: {}.",
                e
            ),
        };
    });

    handle.join().unwrap();

    Ok(())
}

async fn ur_script_generator_server(
    mut requests: impl Stream<Item = ServiceRequest<GenerateURScript::Service>> + Unpin,
    tf_lookup_client: r2r::Client<LookupTransform::Service>,
    templates: tera::Tera,
) -> Result<(), Box<dyn std::error::Error>> {
    let template_names = templates
        .get_template_names()
        .map(|x| x.to_string())
        .collect::<Vec<String>>();
    if template_names.len() == 0 {
        r2r::log_error!(
            "ur_script_generator",
            "Couldn't find any Tera templates."
        );
    }
    for template in &template_names {
        r2r::log_info!(
            "ur_script_generator",
            "Found template: {:?}",
            template
        );
    }

    loop {
        match requests.next().await {
            Some(request) => {
                match template_names.contains(&format!("{}.script", &request.message.command)) {
                    true => match generate_script(&request.message, &tf_lookup_client, &templates).await {
                        Some(script) => {
                            r2r::log_info!(
                                &format!(
                                    "ur_script_generator",
                                ),
                                "Generated UR Script: \n{}",
                                script
                            );
                            request
                                .respond(GenerateURScript::Response {
                                    success: true,
                                    script,
                                })
                                .expect("Could not send service response.")
                        }
                        None => {
                            r2r::log_error!(
                                &format!(
                                    "ur_script_generator",
                                ),
                                "Failed to generate UR Script.",
                            );
                            request
                                .respond(GenerateURScript::Response {
                                    success: false,
                                    script: String::default(),
                                })
                                .expect("Could not send service response.")
                        }
                    },
                    false => {
                        r2r::log_warn!(
                            "ur_script_generator",
                            "Template doesn't exist for command: {}.",
                            request.message.command
                        );
                        request
                            .respond(GenerateURScript::Response {
                                success: false,
                                script: String::default(),
                            })
                            .expect("Could not send service response.")
                    }
                };
            }
            None => break,
        }
    }

    Ok(())
}

async fn generate_script(
    message: &r2r::ur_tools_msgs::srv::GenerateURScript::Request,
    tf_lookup_client: &r2r::Client<LookupTransform::Service>,
    templates: &tera::Tera
) -> Option<String> {
    let empty_context = tera::Context::new();
    let interpreted_message = interpret_message(message, &tf_lookup_client).await;
    if interpreted_message.valid {
        match templates.render(
            &format!("{}.script", message.command),
            match &tera::Context::from_serialize(interpreted_message) {
                Ok(context) => context,
                Err(e) => {
                    r2r::log_error!(
                        "ur_script_generator",
                        "Creating a Tera Context from a serialized Interpretation failed with: {}.",
                        e
                    );
                    r2r::log_warn!(
                        "ur_script_generator",
                        "An empty Tera Context will be used instead."
                    );
                    &empty_context
                }
            },
        ) {
            Ok(script) => Some(script),
            Err(e) => {
                r2r::log_error!(
                    "ur_script_generator",
                    "Rendering the {}.script Tera Template failed with: {}.",
                    message.command,
                    e
                );
                None
            }
        }
    } else {
        None
    }
}

async fn interpret_message(
    message: &r2r::ur_tools_msgs::srv::GenerateURScript::Request,
    tf_lookup_client: &r2r::Client<LookupTransform::Service>,
) -> Interpretation {
    let target_in_base = match message.use_joint_positions && message.command == "move_j" {
        true => pose_to_string(&TransformStamped::default()),
        false => {
            let deadline = match message.tf_lookup_deadline {
                0 => {
                    r2r::log_warn!(
                        "ur_script_generator",
                        "TF Lookup deadline was set to 0ms, changing to 3000ms."
                    );
                    3000
                }
                _ => message.tf_lookup_deadline,
            };
            match lookup_tf(
                BASEFRAME_ID,
                &message.goal_feature_name,
                deadline,
                tf_lookup_client,
            )
            .await
            {
                Some(transform) => pose_to_string(&transform),
                None => "failed".to_string(),
            }
        }
    };

    let tcp_in_faceplate = match message.use_joint_positions && message.command == "move_j" {
        true => pose_to_string(&TransformStamped::default()),
        false => {
            let deadline = match message.tf_lookup_deadline {
                0 => {
                    r2r::log_warn!(
                        "ur_script_generator",
                        "TF Lookup deadline was set to 0ms, changing to 3000ms."
                    );
                    3000
                }
                _ => message.tf_lookup_deadline,
            };
            match lookup_tf(FACEPLATE_ID, &message.tcp_name, deadline, tf_lookup_client).await {
                Some(transform) => pose_to_string(&transform),
                None => "failed".to_string(),
            }
        }
    };

    let valid = tcp_in_faceplate != "failed" && target_in_base != "failed";

    Interpretation {
        valid,
        command: message.command.to_string(),
        acceleration: message.acceleration,
        velocity: message.velocity,
        use_execution_time: message.use_execution_time,
        execution_time: message.execution_time,
        use_blend_radius: message.use_blend_radius,
        blend_radius: message.blend_radius,
        use_joint_positions: message.use_joint_positions,
        joint_positions: joint_pose_to_string(message.joint_positions.clone()),
        use_preferred_joint_config: message.use_preferred_joint_config,
        preferred_joint_config: joint_pose_to_string(message.preferred_joint_config.clone()),
        use_payload: message.use_payload,
        payload: payload_to_string(message.payload.clone()),
        target_in_base,
        tcp_in_faceplate,
    }
}

fn payload_to_string(p: Payload) -> String {
    format!(
        "{},[{},{},{}],[{},{},{},{},{},{}]",
        p.mass, p.cog_x, p.cog_y, p.cog_z, p.ixx, p.iyy, p.izz, p.ixy, p.ixz, p.iyz
    )
}

fn joint_pose_to_string(jp: JointPositions) -> String {
    format!(
        "[{},{},{},{},{},{}]",
        jp.j0, jp.j1, jp.j2, jp.j3, jp.j4, jp.j5
    )
}

fn pose_to_string(tf_stamped: &TransformStamped) -> String {
    let x = tf_stamped.transform.translation.x;
    let y = tf_stamped.transform.translation.y;
    let z = tf_stamped.transform.translation.z;
    let rot = tf_stamped.transform.rotation.clone();
    let angle = 2.0 * rot.w.acos();
    let den = (1.0 - rot.w.powi(2)).sqrt();
    let (rx, ry, rz) = match den < 0.001 {
        true => (rot.x * angle, rot.y * angle, rot.z * angle),
        false => (
            (rot.x / den) * angle,
            (rot.y / den) * angle,
            (rot.z / den) * angle,
        ),
    };
    format!("p[{},{},{},{},{},{}]", x, y, z, rx, ry, rz)
}

async fn lookup_tf(
    parent_id: &str,
    child_id: &str,
    deadline: i32,
    tf_lookup_client: &r2r::Client<LookupTransform::Service>,
) -> Option<TransformStamped> {
    let request = LookupTransform::Request {
        parent_id: parent_id.to_string(),
        child_id: child_id.to_string(),
        deadline,
    };

    let response = tf_lookup_client
        .request(&request)
        .expect("Could not send TF Lookup request.")
        .await
        .expect("Cancelled.");

    r2r::log_info!(
        "ur_script_generator",
        "Request to TF Lookup sent."
    );

    match response.success {
        true => Some(response.transform),
        false => {
            r2r::log_error!(
                "ur_script_generator",
                "Couldn't lookup TF for parent '{}' and child '{}'.",
                parent_id,
                child_id
            );
            None
        }
    }
}
